### PR TITLE
#109: Use rockspec for Lua dependency management.

### LINKED ISSUE #109: Rockspec for row-level-security-lua
## Situation

While users can't directly use `luarocks` inside an Exaol Lua script, it is still beneficial to have a Rockspec in place. The main reason is that building an dependency management are the centralized in the project, which reduces code duplication.

* Abstracts the test execution
* Central point to declare all modules (internal and external dependencies)
* Separate between runtime and test modules

Best case we even find a way to reuse that information to create the all-in-one package with `amalg.lua`.

## Acceptance Criteria

* Rockspec exists
* `luarocks --local make` works
* `luarocks --local test` works
* `make` and `test` used in CI builds
* Either `amalg.lua` is fed with info from the rock spec, or this ticket documents, why that is not possible.

### PR DESCRIPTION
**Pull request by redcatbear**
Closes #109.

### CODE REVIEW

File: .github/workflows/release_droid_prepare_original_checksum.yml [hunk 1]
```diff
@@ -39,8 +36,10 @@ jobs:
       - name: "Prepare container reuse"
         run: |
           echo 'testcontainers.reuse.enable=true' > "/home/runner/.testcontainers.properties"
-      - name: Run tests and build with Maven
-        run: mvn -B -Dorg.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener=warn clean verify --file pom.xml
+      - name: "Build with Maven"
+        run: |
+          eval $(luarocks path)
+          mvn verify -B -Dorg.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener=warn
```
Comment (kaklakariada) [inline .github/workflows/release_droid_prepare_original_checksum.yml:42]:
```suggestion
          mvn verify --batch-mode -Dorg.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener=warn
```

File: .github/workflows/release_droid_print_quick_checksum.yml [hunk 2]
```diff
@@ -5,38 +5,36 @@ on:
...
+      uses: actions/cache@v2
+      with:
+        path: ~/.m2/repository
+        key: ${{ runner.os }}-maven-${{ hashFiles('**/pom.xml') }}
+        restore-keys: |
+          ${{ runner.os }}-maven-
+    - name: "Build with Maven skipping tests"
+      run: |
+        eval $(luarocks path)
+        mvn clean verify -B -DskipTests -Dorg.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener=warn
```
Comment (kaklakariada) [inline .github/workflows/release_droid_print_quick_checksum.yml:37]:
```suggestion
        mvn clean verify --batch-mode -DskipTests -Dorg.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener=warn
```

File: .github/workflows/release_droid_upload_github_release_assets.yml [hunk 3]
```diff
@@ -11,36 +11,39 @@ jobs:
...
             ${{ runner.os }}-maven-
-      - name: Build with Maven skipping tests
-        run: mvn clean verify -DskipTests
+      - name: "Prepare container reuse"
+        run: |
+          echo 'testcontainers.reuse.enable=true' > "/home/runner/.testcontainers.properties"
+      - name: "Build with Maven"
+        run: |
+          eval $(luarocks path)
+          mvn clean verify -B -DskipTests -Dorg.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener=warn
```
Comment (kaklakariada) [inline .github/workflows/release_droid_upload_github_release_assets.yml:46]:
```suggestion
          mvn clean verify --batch-mode -DskipTests -Dorg.slf4j.simpleLogger.log.org.apache.maven.cli.transfer.Slf4jMavenTransferListener=warn
```

File: .github/workflows/ci-build.yml [hunk 4]
```diff
@@ -18,17 +18,7 @@ jobs:
...
-          luarocks install --local LuaUnit
-          luarocks install --local Mockagne
-          luarocks install --local lua-cjson 2.1.0
-          luarocks install --local remotelog
-          luarocks install --local exaerror
-          luarocks install --local virtual-schema-common-lua
-          luarocks install --local amalg
+          luarocks --local install *.rockspec --deps-only
       - name: "Set up JDK"
         uses: actions/setup-java@v2
```
Comment (kaklakariada) [inline .github/workflows/ci-build.yml:23]:
```suggestion
        uses: actions/setup-java@v3
```

File: .github/workflows/release_droid_prepare_original_checksum.yml [hunk 5]
```diff
@@ -7,28 +7,25 @@ jobs:
...
         with:
-          fetch-depth: 0
-      - name: Set up JDK 11
-        uses: actions/setup-java@v1
+          luarocksVersion: "3.8.0"
+      - name: "Install dependencies"
+        run: |
+          luarocks --local install *.rockspec --deps-only
+      - name: "Set up JDK"
+        uses: actions/setup-java@v2
```
Comment (kaklakariada) [inline .github/workflows/release_droid_prepare_original_checksum.yml:25]:
```suggestion
        uses: actions/setup-java@v3
```

File: .github/workflows/release_droid_prepare_original_checksum.yml [hunk 6]
```diff
@@ -7,28 +7,25 @@ jobs:
...
         with:
-          java-version: 11
+          java-version: '11'
+          distribution: 'adopt'
```
Comment (kaklakariada) [inline .github/workflows/release_droid_prepare_original_checksum.yml:28]:
```suggestion
          distribution: 'temurin'
```

File: .github/workflows/release_droid_prepare_original_checksum.yml [hunk 7]
```diff
@@ -7,28 +7,25 @@ jobs:
...
       - name: Cache local Maven repository
         uses: actions/cache@v2
```
Comment (kaklakariada) [inline .github/workflows/release_droid_prepare_original_checksum.yml:30]:
```suggestion
        uses: actions/cache@v3
```

File: .github/workflows/release_droid_print_quick_checksum.yml [hunk 8]
```diff
@@ -5,38 +5,36 @@ on:
 
 jobs:
   build:
-    runs-on: ubuntu-latest
-    steps:
-      - name: Checkout the repository
-        uses: actions/checkout@v2
+    - uses: actions/checkout@v2
```
Comment (kaklakariada) [inline .github/workflows/release_droid_print_quick_checksum.yml:8]:
```suggestion
    - uses: actions/checkout@v3
```

File: .github/workflows/release_droid_print_quick_checksum.yml [hunk 9]
```diff
@@ -5,38 +5,36 @@ on:
...
+      with:
+        luarocksVersion: "3.8.0"
+    - name: "Install dependencies"
+      run: |
+        luarocks --local install *.rockspec --deps-only
+    - name: "Set up JDK"
+      uses: actions/setup-java@v2
+      with:
+        java-version: '11'
+        distribution: 'adopt'
```
Comment (kaklakariada) [inline .github/workflows/release_droid_print_quick_checksum.yml:26]:
```suggestion
    - name: "Set up JDK"
      uses: actions/setup-java@v3
      with:
        java-version: '11'
        distribution: 'temurin'
```

File: .github/workflows/release_droid_print_quick_checksum.yml [hunk 10]
```diff
@@ -5,38 +5,36 @@ on:
...
+    - name: Cache local Maven repository
+      uses: actions/cache@v2
```
Comment (kaklakariada) [inline .github/workflows/release_droid_print_quick_checksum.yml:28]:
```suggestion
      uses: actions/cache@v3
```

File: .github/workflows/release_droid_upload_github_release_assets.yml [hunk 11]
```diff
@@ -11,36 +11,39 @@ jobs:
...
-          sudo luarocks install Mockagne
-          sudo luarocks install lua-cjson
-          sudo luarocks install remotelog
-          sudo luarocks install exaerror
-          sudo luarocks install amalg
-      - name: Set up JDK 11
-        uses: actions/setup-java@v1
+          luarocks --local install *.rockspec --deps-only
+      - name: "Set up JDK"
+        uses: actions/setup-java@v2
```
Comment (kaklakariada) [inline .github/workflows/release_droid_upload_github_release_assets.yml:29]:
```suggestion
        uses: actions/setup-java@v3
```

File: .github/workflows/release_droid_upload_github_release_assets.yml [hunk 12]
```diff
@@ -11,36 +11,39 @@ jobs:
...
         with:
-          java-version: 11
+          java-version: '11'
+          distribution: 'adopt'
```
Comment (kaklakariada) [inline .github/workflows/release_droid_upload_github_release_assets.yml:32]:
```suggestion
          distribution: 'temurin'
```

File: row-level-security-lua-1.1.0-1.rockspec [hunk 13]
```diff
@@ -0,0 +1,63 @@
...
+    "message_expander",
+    -- from virtual-schema-common-lua"
+    "exasolvs.AbstractVirtualSchemaAdapter", "exasolvs.AdapterProperties", "exasolvs.RequestDispatcher",
+    "exasolvs.QueryRenderer", "text"
+}
+
+local item_path_list = ""
+for i = 1, #package_items do
+    item_path_list = item_path_list .. " " .. package_items[i]
+end
```
Comment (kaklakariada) [inline row-level-security-lua-1.1.0-1.rockspec:55]: Can you use `table.concat()` instead?

File: row-level-security-lua-1.1.0-1.rockspec [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+
+build = {
+    type = "command",
+    build_command = "cd " .. src .. " && amalg.lua "
+            .. "-o ../../../target/row-level-security-dist-" .. tag .. ".lua "
+            .. "-s entry.lua"
+            .. item_path_list
```
Comment (kaklakariada) [inline row-level-security-lua-1.1.0-1.rockspec:62]: 👍

**Review by kaklakariada — CHANGES_REQUESTED** [review]

File: row-level-security-lua-1.1.0-1.rockspec (hunk 13 above)
Comment (redcatbear) [reply row-level-security-lua-1.1.0-1.rockspec:55]: Standard library is unfortunately not available in a rockspec.

**Review by kaklakariada — APPROVED** [review]

**Pull request merged**